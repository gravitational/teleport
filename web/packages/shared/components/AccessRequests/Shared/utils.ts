/**
 * Copyright 2024 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { formatDuration, intervalToDuration } from 'date-fns';

import { ResourceMap } from '../NewRequest';

export function getFormattedDurationTxt({
  start,
  end,
}: {
  start: Date;
  end: Date;
}) {
  return formatDuration(intervalToDuration({ start, end }), {
    format: ['weeks', 'days', 'hours', 'minutes'],
  });
}

export function getNumAddedResources(addedResources: ResourceMap) {
  return (
    Object.keys(addedResources.node).length +
    Object.keys(addedResources.db).length +
    Object.keys(addedResources.app).length +
    Object.keys(addedResources.kube_cluster).length +
    Object.keys(addedResources.user_group).length +
    Object.keys(addedResources.windows_desktop).length
  );
}
