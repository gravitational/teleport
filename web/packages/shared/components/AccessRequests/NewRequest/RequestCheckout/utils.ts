/**
 * Teleport
 * Copyright (C) 2024  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { addHours, addDays, isAfter } from 'date-fns';

import { Option } from 'shared/components/Select';

import { getFormattedDurationTxt } from '../../Shared/utils';

// Preset hour options for the time option dropdowns.
export const presetHours = [1, 2, 3, 4, 6, 8, 12, 18];
// Preset day options for the time option dropdowns.
export const presetDays = [1, 2, 3, 4, 5, 6, 7];
const HourInMs = 60 * 60 * 1000;

export function getPendingRequestDurationOptions(
  accessRequestCreated: Date,
  maxDuration: number
): Option<number>[] {
  const createdDate = new Date(accessRequestCreated);

  // Backend limits max pending date to be 1 week.
  // MaxDuration can get greater than 1 week.
  // Pick the smaller of the two.
  let maxPendingDate = new Date(maxDuration);
  const maxPendingDay = presetDays.length - 1;
  const possiblySmallerMaxPendingDate = addDays(
    createdDate,
    presetDays[maxPendingDay]
  );
  if (maxPendingDate.getTime() > possiblySmallerMaxPendingDate.getTime()) {
    maxPendingDate = possiblySmallerMaxPendingDate;
  }

  const createdTimestamp = createdDate.getTime();
  const pendingTimestamp = maxPendingDate.getTime();

  let durationOpts: Option<number>[] = [];

  const totalHoursDiff = (pendingTimestamp - createdTimestamp) / HourInMs;

  // If there is less than an hour available, return
  // it as max duration as the only option.
  if (totalHoursDiff <= 1) {
    durationOpts.push({
      value: pendingTimestamp,
      label: getFormattedDurationTxt({
        start: createdDate,
        end: maxPendingDate,
      }),
    });
    return durationOpts;
  }

  // Add preset hour options up to maximum allowed.
  for (const hour of presetHours) {
    const updatedDateTime = addHours(createdDate, hour);
    if (isAfter(updatedDateTime, maxPendingDate)) {
      break;
    }
    durationOpts.push({
      value: updatedDateTime.getTime(),
      label: getFormattedDurationTxt({
        start: createdDate,
        end: updatedDateTime,
      }),
    });
  }

  // Add preset days up to maximum allowed.
  if (totalHoursDiff >= 24) {
    for (const day of presetDays) {
      const updatedEndDate = addDays(createdDate, day);
      if (isAfter(updatedEndDate, maxPendingDate)) {
        break;
      }
      durationOpts.push({
        value: updatedEndDate.getTime(),
        label: getFormattedDurationTxt({
          start: createdDate,
          end: updatedEndDate,
        }),
      });
    }
  }

  const lastDurationOption = durationOpts[durationOpts.length - 1];
  if (pendingTimestamp > lastDurationOption.value) {
    durationOpts.push({
      value: pendingTimestamp,
      label: getFormattedDurationTxt({
        start: createdDate,
        end: maxPendingDate,
      }),
    });
  }

  return durationOpts;
}

/**
 * Backend expects the maxDuration field to be set to some value
 * which on dry run's will get overwritten to whatever the backend
 * defualt max maxDuration is. Leaving maxDuration field undefined
 * for dry run's, for some reason gets respected.
 */
export function getDryRunMaxDuration() {
  const sevenDaysInMs = 1000 * 60 * 60 * 24 * 7;
  return new Date(Date.now() + sevenDaysInMs);
}
