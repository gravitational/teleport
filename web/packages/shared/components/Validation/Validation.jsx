/*
 * Teleport
 * Copyright (C) 2023  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react';

import { isObject } from 'shared/utils/highbar';

import Logger from '../../libs/logger';

const logger = Logger.create('validation');

// Validator handles input validation
export default class Validator {
  valid = true;

  constructor() {
    // store subscribers
    this._subs = [];
  }

  // adds a callback to the list of subscribers
  subscribe(cb) {
    this._subs.push(cb);
  }

  // removes a callback from the list of subscribers
  unsubscribe(cb) {
    const index = this._subs.indexOf(cb);
    if (index > -1) {
      this._subs.splice(index, 1);
    }
  }

  addResult(result) {
    // result can be a boolean value or an object
    let isValid = false;
    if (isObject(result)) {
      isValid = result.valid;
    } else {
      logger.error(`rule should return a valid object`);
    }

    this.valid = this.valid && Boolean(isValid);
  }

  reset() {
    this.valid = true;
    this.validating = false;
  }

  validate() {
    this.reset();
    this.validating = true;
    this._subs.forEach(cb => {
      try {
        cb();
      } catch (err) {
        logger.error(err);
      }
    });

    return this.valid;
  }
}

const ValidationContext = React.createContext({});

export function Validation(props) {
  const [validator] = React.useState(() => new Validator());
  // handle render functions
  const children =
    typeof props.children === 'function'
      ? props.children({ validator })
      : props.children;

  return (
    <ValidationContext.Provider value={validator}>
      {children}
    </ValidationContext.Provider>
  );
}

export function useValidation() {
  const value = React.useContext(ValidationContext);
  if (!(value instanceof Validator)) {
    logger.warn('Missing Validation Context declaration');
  }

  return value;
}
