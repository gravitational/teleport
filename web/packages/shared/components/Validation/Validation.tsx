/*
 * Teleport
 * Copyright (C) 2023  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react';

import { Logger } from 'design/logger';

import { isObject } from 'shared/utils/highbar';
import { Store, useStore } from 'shared/libs/stores';

import { ValidationResult } from './rules';

const logger = new Logger('validation');

/** A per-rule callback that will be executed during validation. */
type RuleCallback = () => void;

export type Result = ValidationResult | boolean;

type ValidatorState = {
  /** Indicates whether the last validation was successful. */
  valid: boolean;
  /**
   * Indicates whether the validator has been activated by a call to
   * `validate`.
   */
  validating: boolean;
};

/** A store that handles input validation and makes its results accessible. */
export default class Validator extends Store<ValidatorState> {
  state = {
    valid: true,
    validating: false,
  };

  /**
   * @deprecated For temporary Enterprise compatibility only. Use {@link state}
   * instead.
   */
  valid = true;

  /** Callbacks that will be executed upon validation. */
  private ruleCallbacks: RuleCallback[] = [];

  /** Adds a rule callback that will be executed upon validation. */
  addRuleCallback(cb: RuleCallback) {
    this.ruleCallbacks.push(cb);
  }

  /** Removes a rule callback. */
  removeRuleCallback(cb: RuleCallback) {
    const index = this.ruleCallbacks.indexOf(cb);
    if (index > -1) {
      this.ruleCallbacks.splice(index, 1);
    }
  }

  addResult(result: Result) {
    // result can be a boolean value or an object
    let isValid = false;
    if (isObject(result)) {
      isValid = result.valid;
    } else {
      logger.error(`rule should return a valid object`);
    }

    const valid = this.state.valid && Boolean(isValid);
    this.setState({ valid });
    this.valid = valid;
  }

  reset() {
    this.setState({
      valid: true,
      validating: false,
    });
    this.valid = true;
  }

  validate() {
    this.reset();
    this.setState({ validating: true });
    for (const cb of this.ruleCallbacks) {
      try {
        cb();
      } catch (err) {
        logger.error(err);
      }
    }

    return this.state.valid;
  }
}

const ValidationContext = React.createContext<Validator | undefined>(undefined);

type ValidationRenderFunction = (arg: {
  validator: Validator;
}) => React.ReactNode;

/**
 * Installs a validation context that provides a {@link Validator} store. The
 * store can be retrieved either through {@link useValidation} hook or by a
 * render callback, e.g.:
 *
 * ```
 * function Component() {
 *   return (
 *     <Validation>
 *       {({validator}) => (
 *         <>
 *           (...)
 *           <button onClick={() => validator.validate()}>Validate</button>
 *         </>
 *       )}
 *     </Validation>
 *   );
 * }
 * ```
 *
 * The simplest way to use validation is validating on the view layer: just use
 * a `rule` prop with `FieldInput` or a similar component and pass a rule like
 * `requiredField`.
 *
 * Unfortunately, due to architectural limitations, this will not work well in
 * scenarios where information about validity about given field or group of
 * fields is required outside that field. In cases like this, the best option
 * is to validate the model during render time on the top level (for example,
 * execute an entire set of rules on a model using `runRules`). The result of
 * model validation will then contain information about the validity of each
 * field. It can then be used wherever it's needed, and also attached to
 * appropriate inputs with a `precomputed` validation rule. Example:
 *
 * ```
 * function Component(model: Model) {
 *   const rules = {
 *     name: requiredField('required'),
 *     email: requiredEmailLike,
 *   }
 *   const validationResult = runRules(model, rules);
 * }
 * ```
 *
 * Note that, as this example shows clearly, the validator itself, despite its
 * name, doesn't really validate anything -- it merely aggregates validation
 * results. Also it's worth mentioning that the validator will not do it
 * without our help -- each validated field needs to be actually attached to a
 * field, even if using a `precomputed` rule, for this to work. The validation
 * callbacks registered by validation rules on the particular fields are the
 * actual points where the errors are consumed by the validator.
 */
export function Validation(props: {
  children?: React.ReactNode | ValidationRenderFunction;
}) {
  const [validator] = React.useState(() => new Validator());
  useStore(validator);
  // handle render functions
  const children =
    typeof props.children === 'function'
      ? props.children({ validator })
      : props.children;

  return (
    <ValidationContext.Provider value={validator}>
      {children}
    </ValidationContext.Provider>
  );
}

export function useValidation(): Validator {
  const validator = React.useContext(ValidationContext);
  if (!validator) {
    throw new Error('useValidation() called without a validation context');
  }
  return useStore(validator);
}
