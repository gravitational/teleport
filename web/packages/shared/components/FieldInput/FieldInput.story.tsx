/**
 * Teleport
 * Copyright (C) 2023  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { ButtonPrimary, Text } from 'design';
import { EmailSolid } from 'design/Icon';

import Validation from '../../components/Validation';
import { requiredEmailLike, requiredField } from '../Validation/rules';
import FieldInput from './FieldInput';

export default {
  title: 'Shared',
};

export const Fields = () => (
  <Validation>
    {({ validator }) => (
      <>
        <FieldInput
          label="Label"
          helperText="Optional helper text"
          name="optional name"
          onChange={() => {}}
          value={'value'}
          icon={EmailSolid}
          size="large"
          rule={requiredEmailLike}
        />
        <FieldInput
          label="Label with placeholder"
          name="optional name"
          onChange={() => {}}
          placeholder="placeholder"
          value={''}
        />
        <FieldInput
          label="Label with tooltip"
          name="optional name"
          onChange={() => {}}
          placeholder="placeholder"
          value={''}
          toolTipContent={<Text>Hello world</Text>}
        />
        <FieldInput
          label="Label with helper text and tooltip"
          helperText="Helper text"
          toolTipContent={<Text>Hello world</Text>}
          name="optional name"
          onChange={() => {}}
          placeholder="placeholder"
          value={''}
        />
        <FieldInput placeholder="without label" onChange={() => {}} />
        <FieldInput
          label="Required"
          rule={requiredField('So required. Much mandatory.')}
          onChange={() => {}}
          value=""
        />
        <ButtonPrimary onClick={() => validator.validate()}>
          Validate
        </ButtonPrimary>
      </>
    )}
  </Validation>
);

Fields.storyName = 'FieldInput';
