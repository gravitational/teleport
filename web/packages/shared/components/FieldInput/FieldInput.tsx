/**
 * Teleport
 * Copyright (C) 2023  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { forwardRef } from 'react';
import { Box, Input, LabelInput, Text } from 'design';

import { ToolTipInfo } from 'shared/components/ToolTip';
import { useRule } from 'shared/components/Validation';

const FieldInput = forwardRef<HTMLInputElement, FieldInputProps>(
  (
    {
      label,
      labelTip,
      value,
      onChange,
      onKeyPress,
      placeholder,
      defaultValue,
      min,
      max,
      rule = defaultRule,
      name,
      type = 'text',
      autoFocus = false,
      autoComplete = 'off',
      inputMode = 'text',
      readonly = false,
      toolTipContent = null,
      disabled = false,
      markAsError = false,
      ...styles
    },
    ref
  ) => {
    const { valid, message } = useRule(rule(value));
    const hasError = !valid;
    const labelText = hasError ? message : label;
    const $inputElement = (
      <Input
        mt={1}
        ref={ref}
        type={type}
        name={name}
        hasError={hasError || markAsError}
        placeholder={placeholder}
        autoFocus={autoFocus}
        value={value}
        min={min}
        max={max}
        autoComplete={autoComplete}
        onChange={onChange}
        onKeyPress={onKeyPress}
        readOnly={readonly}
        inputMode={inputMode}
        defaultValue={defaultValue}
        disabled={disabled}
      />
    );

    return (
      <Box mb="4" {...styles}>
        {label ? (
          <LabelInput mb={0} hasError={hasError}>
            {toolTipContent ? (
              <>
                <span
                  css={{
                    marginRight: '4px',
                    verticalAlign: 'middle',
                  }}
                >
                  {labelText}
                  {labelTip && <LabelTip text={labelTip} />}
                </span>
                <ToolTipInfo children={toolTipContent} />
              </>
            ) : (
              <>
                {labelText}
                {labelTip && <LabelTip text={labelTip} />}
              </>
            )}
            {$inputElement}
          </LabelInput>
        ) : (
          $inputElement
        )}
      </Box>
    );
  }
);

const defaultRule = () => () => ({ valid: true });

const LabelTip = ({ text }) => (
  <Text as="span" style={{ fontWeight: 'normal' }}>{` - ${text}`}</Text>
);

export default FieldInput;

export type FieldInputProps = {
  value?: string;
  label?: string;
  labelTip?: string;
  placeholder?: string;
  autoFocus?: boolean;
  autoComplete?: 'off' | 'on' | 'one-time-code';
  type?: 'email' | 'text' | 'password' | 'number' | 'date' | 'week';
  inputMode?: 'text' | 'numeric';
  rule?: (options: unknown) => () => unknown;
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  onKeyPress?: (e: React.KeyboardEvent<HTMLInputElement>) => void;
  readonly?: boolean;
  defaultValue?: string;
  min?: number;
  max?: number;
  toolTipContent?: React.ReactNode;
  disabled?: boolean;
  // markAsError is a flag to highlight an
  // input box as error color before validator
  // runs (which marks it as error)
  markAsError?: boolean;
  // TS: temporary handles ...styles
  [key: string]: any;
};
