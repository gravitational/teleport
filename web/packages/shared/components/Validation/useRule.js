/*
 * Teleport
 * Copyright (C) 2023  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react';

import Logger from '../../libs/logger';

import { useValidation } from './Validation';

const logger = Logger.create('validation');

/**
 * useRule subscribes to validation requests upon which executes validate() callback
 */
export default function useRule(cb) {
  if (typeof cb !== 'function') {
    logger.warn(`useRule(fn), fn() must be a function`);
    return;
  }

  const [, rerender] = React.useState();
  const validator = useValidation();

  // register to validation context to be called on cb()
  React.useEffect(() => {
    function onValidate() {
      if (validator.state.validating) {
        const result = cb();
        validator.addResult(result);
        rerender({});
      }
    }

    // subscribe to store changes
    validator.addRuleCallback(onValidate);

    // unsubscribe on unmount
    function cleanup() {
      validator.removeRuleCallback(onValidate);
    }

    return cleanup;
  }, [cb]);

  // if validation has been requested, cb right away.
  if (validator.state.validating) {
    return cb();
  }

  return { valid: true };
}
