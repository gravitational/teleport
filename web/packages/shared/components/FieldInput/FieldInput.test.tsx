/**
 * Teleport
 * Copyright (C) 2023  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react';
import { screen } from '@testing-library/react';

import { render, fireEvent } from 'design/utils/testing';

import { darkTheme } from 'design/theme';

import * as useRule from '../Validation/useRule';

import FieldInput from './FieldInput';

test('valid values, autofocus, onChange, onKeyPress', () => {
  const rule = jest.fn();
  const onChange = jest.fn();
  const onKeyPress = jest.fn();

  // mock positive validation
  jest.spyOn(useRule, 'default').mockReturnValue({ valid: true, message: '' });

  render(
    <FieldInput
      placeholder="placeholderText"
      autoFocus={true}
      label="labelText"
      rule={rule}
      onChange={onChange}
      onKeyPress={onKeyPress}
    />
  );

  // test label is displayed
  expect(screen.getByText('labelText')).toBeInTheDocument();

  // test autofocus prop is respected
  const inputEl = screen.getByPlaceholderText('placeholderText');
  expect(inputEl).toHaveFocus();

  // test onChange prop is respected
  fireEvent.change(inputEl, { target: { value: 'test' } });
  expect(onChange).toHaveBeenCalledTimes(1);

  // test onKeyPress prop is respected
  fireEvent.keyPress(inputEl, { key: 'Enter', keyCode: 13 });
  expect(onKeyPress).toHaveBeenCalledTimes(1);
});

test('input validation error state', () => {
  const rule = jest.fn();
  const errorColor = darkTheme.colors.error.main;

  // mock negative validation
  jest
    .spyOn(useRule, 'default')
    .mockReturnValue({ valid: false, message: 'errorMsg' });

  render(
    <FieldInput
      placeholder="placeholderText"
      label="labelText"
      rule={rule}
      onChange={jest.fn()}
    />
  );

  // test !valid values renders with error message
  const labelEl = screen.getByText('errorMsg');
  expect(labelEl).toHaveStyle({ color: errorColor });

  // test !valid values renders error colors
  const inputEl = screen.getByPlaceholderText('placeholderText');
  expect(inputEl).toHaveStyle({
    'border-color': errorColor,
  });
});
