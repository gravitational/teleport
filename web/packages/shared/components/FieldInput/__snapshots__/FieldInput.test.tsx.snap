// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests 1`] = `
.c0 {
  box-sizing: border-box;
  margin-bottom: 40px;
}

.c6 {
  display: inline-flex;
  align-items: center;
  justify-content: center;
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  margin: 0px;
}

.c3 {
  appearance: none;
  border: 1px solid rgba(255,255,255,0.54);
  border-radius: 4px;
  box-sizing: border-box;
  display: block;
  height: 40px;
  font-size: 16px;
  padding: 0 16px;
  outline: none;
  width: 100%;
  background: #222C59;
  color: #FFFFFF;
  margin-top: 4px;
  border: 2px solid #FF6257;
  padding: 10px 14px;
}

.c3:hover,
.c3:focus,
.c3:active {
  border: 1px solid rgba(255,255,255,0.72);
}

.c3::-ms-clear {
  display: none;
}

.c3::placeholder {
  color: rgba(255,255,255,0.54);
  opacity: 1;
}

.c3:read-only {
  cursor: not-allowed;
}

.c3:disabled {
  color: rgba(255,255,255,0.36);
  border-color: rgba(255,255,255,0.36);
}

.c3:hover,
.c3:focus {
  border: 2px solid #FF6257;
}

.c1 {
  color: #FF6257;
  display: block;
  font-size: 12px;
  width: 100%;
  margin-bottom: 0px;
}

.c7 {
  height: 18px;
  width: 18px;
  color: inherit;
}

.c5 {
  vertical-align: middle;
  display: inline-block;
  height: 18px;
}

.c5:hover {
  cursor: pointer;
}

.c4 {
  margin-right: 4px;
  vertical-align: middle;
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1"
    >
      errorMsg
      <span
        class="c2"
        style="font-weight: normal;"
      >
         - Optional tabel tip
      </span>
      <input
        autocomplete="off"
        class="c3"
        inputmode="text"
        name="optional name"
        type="text"
        value="value"
      />
    </label>
  </div>
  <div
    class="c0"
  >
    <label
      class="c1"
    >
      errorMsg
      <input
        autocomplete="off"
        class="c3"
        inputmode="text"
        name="optional name"
        placeholder="placeholder"
        type="text"
        value=""
      />
    </label>
  </div>
  <div
    class="c0"
  >
    <label
      class="c1"
    >
      <span
        class="c4"
      >
        errorMsg
      </span>
      <span
        class="c5"
        role="icon"
      >
        <span
          class="c6 c7"
        >
          <svg
            fill="currentColor"
            height="20"
            viewBox="0 0 24 24"
            width="20"
          >
            <path
              d="M11.625 8.8125C12.1428 8.8125 12.5625 8.39277 12.5625 7.875C12.5625 7.35723 12.1428 6.9375 11.625 6.9375C11.1072 6.9375 10.6875 7.35723 10.6875 7.875C10.6875 8.39277 11.1072 8.8125 11.625 8.8125Z"
            />
            <path
              d="M10.5 11.25C10.5 10.8358 10.8358 10.5 11.25 10.5C11.6478 10.5 12.0294 10.658 12.3107 10.9393C12.592 11.2206 12.75 11.6022 12.75 12V15.75C13.1642 15.75 13.5 16.0858 13.5 16.5C13.5 16.9142 13.1642 17.25 12.75 17.25C12.3522 17.25 11.9706 17.092 11.6893 16.8107C11.408 16.5294 11.25 16.1478 11.25 15.75V12C10.8358 12 10.5 11.6642 10.5 11.25Z"
            />
            <path
              clip-rule="evenodd"
              d="M12 2.25C6.61522 2.25 2.25 6.61522 2.25 12C2.25 17.3848 6.61522 21.75 12 21.75C17.3848 21.75 21.75 17.3848 21.75 12C21.75 6.61522 17.3848 2.25 12 2.25ZM3.75 12C3.75 7.44365 7.44365 3.75 12 3.75C16.5563 3.75 20.25 7.44365 20.25 12C20.25 16.5563 16.5563 20.25 12 20.25C7.44365 20.25 3.75 16.5563 3.75 12Z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </span>
      <input
        autocomplete="off"
        class="c3"
        inputmode="text"
        name="optional name"
        placeholder="placeholder"
        type="text"
        value=""
      />
    </label>
  </div>
  <div
    class="c0"
  >
    <label
      class="c1"
    >
      <span
        class="c4"
      >
        errorMsg
        <span
          class="c2"
          style="font-weight: normal;"
        >
           - the label tip
        </span>
      </span>
      <span
        class="c5"
        role="icon"
      >
        <span
          class="c6 c7"
        >
          <svg
            fill="currentColor"
            height="20"
            viewBox="0 0 24 24"
            width="20"
          >
            <path
              d="M11.625 8.8125C12.1428 8.8125 12.5625 8.39277 12.5625 7.875C12.5625 7.35723 12.1428 6.9375 11.625 6.9375C11.1072 6.9375 10.6875 7.35723 10.6875 7.875C10.6875 8.39277 11.1072 8.8125 11.625 8.8125Z"
            />
            <path
              d="M10.5 11.25C10.5 10.8358 10.8358 10.5 11.25 10.5C11.6478 10.5 12.0294 10.658 12.3107 10.9393C12.592 11.2206 12.75 11.6022 12.75 12V15.75C13.1642 15.75 13.5 16.0858 13.5 16.5C13.5 16.9142 13.1642 17.25 12.75 17.25C12.3522 17.25 11.9706 17.092 11.6893 16.8107C11.408 16.5294 11.25 16.1478 11.25 15.75V12C10.8358 12 10.5 11.6642 10.5 11.25Z"
            />
            <path
              clip-rule="evenodd"
              d="M12 2.25C6.61522 2.25 2.25 6.61522 2.25 12C2.25 17.3848 6.61522 21.75 12 21.75C17.3848 21.75 21.75 17.3848 21.75 12C21.75 6.61522 17.3848 2.25 12 2.25ZM3.75 12C3.75 7.44365 7.44365 3.75 12 3.75C16.5563 3.75 20.25 7.44365 20.25 12C20.25 16.5563 16.5563 20.25 12 20.25C7.44365 20.25 3.75 16.5563 3.75 12Z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </span>
      <input
        autocomplete="off"
        class="c3"
        inputmode="text"
        name="optional name"
        placeholder="placeholder"
        type="text"
        value=""
      />
    </label>
  </div>
  <div
    class="c0"
  >
    <input
      autocomplete="off"
      class="c3"
      inputmode="text"
      placeholder="without label"
      type="text"
      value=""
    />
  </div>
</div>
`;
